jdbc/src/main/java/nextstep/jdbc/JdbcTemplate.java

```diff
@@ -1,9 +1,100 @@
+import javax.sql.DataSource;
+import nextstep.exception.DataAccessException;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
 public class JdbcTemplate {
 
     private static final Logger log = LoggerFactory.getLogger(JdbcTemplate.class);
+
+    private final DataSource dataSource;
+
+    public JdbcTemplate(DataSource dataSource) {
+        this.dataSource = dataSource;
+    }
+
+    public void update(String sql, Object... args) {
+        update(sql, createPreparedStatementSetter(args));
+    }
+
+    public void update(String sql, PreparedStatementSetter pss) {
+        execute(sql, pss, PreparedStatement::executeUpdate);
+    }
+
+    public <T> List<T> queryForList(String sql, Class<T> type, Object... args) {
+        return query(sql, (rs, rowNum) -> getSingleObject(rs, type), createPreparedStatementSetter(args));
+    }
+
+    public <T> Optional<T> queryForObject(String sql, Class<T> type, Object... args) {
+        return queryForObject(sql, (rs, rowNum) -> getSingleObject(rs, type), args);
+    }
+
+    public <T> Optional<T> queryForObject(String sql, RowMapper<T> rowMapper, Object... args) {
+        List<T> result = query(sql, new RowMapperExtractor<>(rowMapper), createPreparedStatementSetter(args));
+
+        if (result.isEmpty()) {
+            throw new DataAccessException("조회된 데이터가 없습니다.");
+        }
+        if (result.size() > 1) {
+            throw new DataAccessException("조회된 데이터가 2개 이상입니다.");
+        }
+
+        return Optional.ofNullable(result.get(0));
+    }
+
+    public <T> List<T> query(String sql, RowMapper<T> rowMapper, PreparedStatementSetter pss) {
+        return query(sql, new RowMapperExtractor<>(rowMapper), pss);
+    }
+
+    public <T> List<T> query(String sql, RowMapperExtractor<T> extractor, PreparedStatementSetter pss) {
+        return execute(sql, pss, pstmt -> {
+            try (ResultSet rs = pstmt.executeQuery()) {
+                return extractor.extractData(rs);
+            }
+        });
+    }
+
+    private <T> T execute(String sql, PreparedStatementSetter pss, PreparedStatementCallback<T> action) {
+        try (Connection conn = dataSource.getConnection();
+            PreparedStatement pstmt = conn.prepareStatement(sql)) {
+            log.debug("query : {}", sql);
+
+            pss.setValue(pstmt);
+            return action.doInPreparedStatement(pstmt);
+        } catch (SQLException e) {
+            log.error(e.getMessage(), e);
+            throw new DataAccessException(e);
+        }
+    }
+
+    private <T> T getSingleObject(ResultSet rs, Class<T> type) throws SQLException {
+        try {
+            T object = type.getConstructor().newInstance();
+            for (Field field : object.getClass().getDeclaredFields()) {
+                field.setAccessible(true);
+                field.set(object, rs.getObject(field.getName()));
+            }
+            return object;
+        } catch (Exception e) {
+            throw new SQLException();
```

Comment:
1. SQLException에 e를 인자로 넘겨줍시다~~
  ```suggestion
              throw new SQLException(e);
  ```
  기존 로직 상에서 throw 해주었을 때 스택트레이스 입니다.
  ![image](https://user-images.githubusercontent.com/51393021/134905024-2a05410f-6fb0-4a61-825b-c31371320d9a.png)
  
  `suggestion`이후의 스택트레이스 입니다.
  ![image](https://user-images.githubusercontent.com/51393021/134904718-385be543-401f-4e92-9c7f-5c01a955370e.png)
  - 다른 코드에선 인자로 e를 넘겨준걸로 보아 깜박한거라고 생각되지만~~
2. Object 파싱과정에서 나는 에러인데, SQL Exception을 되던지는건 잘 못된 신호를 줄 수 있다고 생각합니다. 어떻게 생각하시나요? ㅎ_ㅎ